src/js/common/utils/accessible-menus.js

```diff
@@ -0,0 +1,349 @@
+const LEFT_ARROW = 37;
+const UP_ARROW = 38;
+const RIGHT_ARROW = 39;
+const DOWN_ARROW = 40;
+const ESCAPE = 27;
+
+function isWideScreen() {
+  return matchMedia('(min-width: 768px)').matches;
+}
+
+
+/**
+ * Finds the closest ancestor for which the callback returns true.
+ *
+ * @param {HTMLElement}  element
+ * @param {function}     testCB   When this returns true on an ancestor, that ancestor is returned
+ * @return {HTMLElement|null}     If there is no ansestor for which the test returns true, null is returned
+ */
+function findNearestAncestor(element, testCB) {
+  const parent = element.parentElement;
+  if (!parent) {
+    return null;
+  }
+  const returnVal = testCB(parent) ? parent : findNearestAncestor(parent, testCB);
+  return returnVal;
+}
+
+
+// Reduce code duplication
+function findNearestLi(element) {
+  return findNearestAncestor(element, el => el.tagName.toLowerCase() === 'li');
+}
+
+
+/**
+ * Returns whether the HTMLElement passed is a menu button. This is only true if:
+ *  1. The element is a button or has [role="button"]
+ *  2. The element has popup
+ *  3. The element has aria-controls that contains the id of an element with [role="menu"]
+ *
+ * @param {HTMLElement} element  The element in question
+ * @return {bool}
+ */
+function isMenuButton(element) {
+  // There is no element.getElementById() :matrix:
+  const menuElement = element.parentElement.querySelector(`#${element.getAttribute('aria-controls')}`);
+  if (!menuElement) {
+    return false;
+  }
+
+  const menuRole = (menuElement.getAttribute('role') || '').toLowerCase();
+
+  return (element.tagName.toLowerCase() === 'button' || element.getAttribute('role').toLowerCase() === 'button') &&
+    ['menu', 'true'].includes(element.getAttribute('aria-haspopup')) &&
+    // The menuElement either is or contains a menu element
+    !!(menuRole === 'menu' || menuElement.querySelector('[role="menu"]'));
```

Comment:
This is fine, but IMO it might help readability if the conditions were split up and assigned to constants or one-line functions that described what's being tested so that this big check reads more naturally